components/gitpod-db/src/long-running-migration/long-running-migration.ts

```diff
@@ -0,0 +1,82 @@
+/**
+ * Copyright (c) 2023 Gitpod GmbH. All rights reserved.
+ * Licensed under the GNU Affero General Public License (AGPL).
+ * See License.AGPL.txt in the project root for license information.
+ */
+
+import { log } from "@gitpod/gitpod-protocol/lib/util/logging";
+import { inject, injectable, multiInject } from "inversify";
+import { DBLongRunningMigration } from "../typeorm/entity/db-long-running-migration";
+import { TypeORM } from "../typeorm/typeorm";
+
+export const LongRunningMigration = Symbol("LongRunningMigration");
+export interface LongRunningMigration {
+    /**
+     * The name of the migration. This is used to store the completion state in the database.
+     * Changing the name will cause the migration to be run again.
+     */
+    getName(): string;
+
+    /**
+     * Returns true if the migration has completed.
+     */
+    runMigrationBatch(): Promise<boolean>;
+}
+
+/**
+ * This service runs all registered long running migrations in batches until they are completed.
+ * It stores the completion state in the database.
+ */
+@injectable()
+export class LongRunningMigrationService {
+    private typeorm: TypeORM;
+    private migrations: LongRunningMigration[];
+
+    public constructor(
+        @inject(TypeORM) typeorm: TypeORM,
+        @multiInject(LongRunningMigration) migrations: LongRunningMigration[],
+    ) {
+        this.typeorm = typeorm;
+        this.migrations = migrations;
```

Comment:
How do we ensure that they are executed in-order? That is if there are 2 migrations injected, how do we make sure the earlier created one executes first?

Reply:
One way to do this, is not inject them, but keep them in a List which naturally provides the ordering of them. Then they also do not need to be injected.

Reply:
We don't have that requirement now, but we could introduce an optional `dependsOn` property. That said I would not make them dependent on each other on a per batch basis, but only such as "x" must be completed before starting me.

Reply:
But we actually need to make sure that not two batch executions run in parallel in case the quorum changed. I'll wrap the process in a distributed lock to ensure that.

Reply:
We already loop over `this.migrations` so if we guarantee that `this.migrations` is stable sorted according to some attribute, or naturally ordered, then we solve the ordering problem.

Reply:
But there is no problem with arbitrary ordering. I wouldn't want us to rely on the order of how batches are running as it would imply that we implicitly depend on other migrations.

Reply:
It means that every long-running migration in the future now needs to explicitly know about the previous migrations, and I'm not sure that's intuitive, especially with respect to how the SQL migrations work.

For example, if I add a migration which binds a project to an organization, I want to know that the migration which migrated all workspaces to organizations has in fact run.

Up to you if you choose to handle. If not, it would be good to comment that the migrations in that list will execute in arbitrary order.

Reply:
We can handle it when we need it, but as I said before ordering the list will not solve this, because the same migration is executed in multiple batches (and could run over weeks). What you are asking for is a dependency to the `completed` state of a certain migration. I think an explicit "depends" property would be better. But let's introduce it when we need it.